src/main/java/com/atlassian/plugins/confluence/markdown/ccma/PageRestrictionService.java

```diff
@@ -0,0 +1,71 @@
+package com.atlassian.plugins.confluence.markdown.ccma;
+
+import com.atlassian.confluence.api.model.content.Content;
+import com.atlassian.confluence.api.model.pagination.PageResponse;
+import com.atlassian.confluence.api.model.people.Subject;
+import com.atlassian.confluence.api.model.people.SubjectType;
+import com.atlassian.confluence.api.model.people.User;
+import com.atlassian.confluence.api.model.permissions.ContentRestriction;
+import com.atlassian.confluence.api.model.permissions.OperationKey;
+import com.atlassian.confluence.user.UserAccessor;
+import com.atlassian.plugin.spring.scanner.annotation.imports.ConfluenceImport;
+import com.atlassian.sal.api.user.UserKey;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Named
+class PageRestrictionService extends PermissionService<Content, String> {
+
+    @Inject
+    PageRestrictionService(@ConfluenceImport UserAccessor userAccessor) {
+        super(userAccessor);
+    }
+
+    @Override
+    String key(Content page) {
+        return page.getId().serialise();
+    }
+
+    @Override
+    List<Content> convert(Collection<PageData> pageDataList) {
+        return pageDataList
+                .stream()
+                .map(PageData::getPage)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    Set<String> getEditGroups(Content page) {
+        final List<Subject> groupSubject = Optional
+                .ofNullable(page.getRestrictions())
+                .map(restrictions -> restrictions.get(OperationKey.UPDATE))
+                .map(ContentRestriction::getRestrictions)
+                .map(restrictions -> restrictions.get(SubjectType.GROUP))
+                .map(PageResponse::getResults)
+                .orElse(Collections.emptyList());
+
+        return groupSubject.stream()
+                .map(com.atlassian.confluence.api.model.people.Group.class::cast)
+                .map(com.atlassian.confluence.api.model.people.Group::getName)
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    Set<UserKey> getEditUsers(Content page) {
+        final List<Subject> userSubjects = Optional.ofNullable(page.getRestrictions())
```

Comment:
In large instances, we may have a thousand users or so. If we're running this migration in the background, then I think it's fine, otherwise, should we consider limit the result set?

Reply:
Yes migration is run in the background. Actually getting individual users settings shouldn't be a heavy task in most case because these settings is "individual" that means user is added in one by one basic manually.

But query all users of all granted group may be heavy in the number of member in the group is big.

Reply:
Let me update the code when picking use from group. Get all users of a group then just pick one of them is way ineffective.

Reply:
Group users fetching logic has been updated to work more effective:
- Only fetch group users if needed
- When fetching group users, only fetch with limited results. It is done by using `Stream.findFirst()` on `UserAccessor.getMembers(group)` iterator which will only fetch users lazily on demand.
- Group members and user group names are now cached (for each run of migration)